packages/flutter_tools/test/integration.shard/hot_reload_websocket_test.dart

```diff
@@ -0,0 +1,193 @@
+  group('WebSocket DWDS connection'
+      '${additionalCommandArgs.isEmpty ? '' : ' with args: $additionalCommandArgs'}', () {
+    // Test configuration constants
+    const debugUrlTimeout = Duration(seconds: 20);
+    const appStartTimeout = Duration(seconds: 15);
+    const hotReloadTimeout = Duration(seconds: 10);
+
+    late Directory tempDir;
+    final project = HotReloadProject();
+    late FlutterRunTestDriver flutter;
+
+    setUp(() async {
+      tempDir = createResolvedTempDirectorySync('hot_reload_websocket_test.');
+      await project.setUpIn(tempDir);
+      flutter = FlutterRunTestDriver(tempDir);
+    });
+
+    tearDown(() async {
+      await flutter.stop();
+      tryToDelete(tempDir);
+    });
+
+    testWithoutContext(
+      'hot reload with headless Chrome WebSocket connection',
+      () async {
+        debugPrint('Starting WebSocket DWDS test with headless Chrome...');
+
+        // Set up listening for app output before starting
+        final stdout = StringBuffer();
+        final sawDebugUrl = Completer<String>();
+        final StreamSubscription<String> subscription = flutter.stdout.listen((String e) {
+          stdout.writeln(e);
+          // Extract the debug connection URL
+          if (e.contains('Waiting for connection from Dart debug extension at http://')) {
+            final debugUrlPattern = RegExp(
+              r'Waiting for connection from Dart debug extension at (http://[^\s]+)',
+            );
+            final Match? match = debugUrlPattern.firstMatch(e);
+            if (match != null && !sawDebugUrl.isCompleted) {
+              sawDebugUrl.complete(match.group(1)!);
+            }
+          }
+        });
+
+        io.Process? chromeProcess;
+        try {
+          // Step 1: Start Flutter app with web-server device (will wait for debug connection)
+          debugPrint('Step 1: Starting Flutter app with web-server device...');
+          // Start the app but don't wait for it to complete - it won't complete until Chrome connects
+          final Future<void> appStartFuture = runFlutterWithWebServerDevice(
+            flutter,
+            additionalCommandArgs: [...additionalCommandArgs, '--no-web-resources-cdn'],
+          );
+
+          // Step 2: Wait for DWDS debug URL to be available
+          debugPrint('Step 2: Waiting for DWDS debug service URL...');
+          final String debugUrl = await sawDebugUrl.future.timeout(
+            debugUrlTimeout,
+            onTimeout: () {
+              throw Exception('DWDS debug URL not found - app may not have started correctly');
+            },
+          );
+          debugPrint('✓ DWDS debug service available at: $debugUrl');
+
+          // Step 3: Launch headless Chrome to connect to DWDS
+          debugPrint('Step 3: Launching headless Chrome to connect to DWDS...');
+          chromeProcess = await _launchHeadlessChrome(debugUrl);
+          debugPrint('✓ Headless Chrome launched and connecting to DWDS');
```

Comment:
Should we remove this part of the test now? It seems like we don't need another chrome session involved since we removed the race condition?

Reply:
No, this is still required to actually initialize DWDS for the application. Breaking the deadlock just fixed the `flutter drive` case where the test driver couldn't start until the service was available, but the service couldn't become available until an initial connection had been made to the server.

Reply:
OK thanks. IIUC this test simulates `flutter run -d web-server` which doesn't launch a browser itself so instead the test launches a headless chrome instance to connect to it and trigger the machinery to start the application.